crdt.go

```diff
@@ -77,19 +77,30 @@ type SessionDAGService interface {
 type Options struct {
 	Logger              logging.StandardLogger
 	RebroadcastInterval time.Duration
-	// The PutHook function is triggered whenever an element
+	// The PutPreHook function is triggered before an element
+	// is added to the datastore (either by a local or remote update),
+	// and only when that addition is considered the prevalent value.
+	PutPreHook func(k ds.Key, v []byte) error
+	// The PutPostHook function is triggered whenever an element
 	// is successfully added to the datastore (either by a local
 	// or remote update), and only when that addition is considered the
 	// prevalent value.
-	PutHook func(k ds.Key, v []byte)
-	// The DeleteHook function is triggered whenever a version of an
+	PutPostHook func(k ds.Key, v []byte)
+	// The DeletePreHook function is triggered before an element is removed
+	// from the datastore (either by a local or remote update).
+	// Unordered and concurrent updates may result in the DeletePreHook being
+	// triggered even though the element is still present in the datastore
+	// because it was re-added. If that is relevant, use Has() to check if the
+	// removed element is still part of the datastore.
+	DeletePreHook func(k ds.Key) error
```

Comment:
Explanation here is weird. I would expect that when a delete happens and a pre-hook is triggered, the element is still in the datastore since it is a PRE-hook. I kind of get what it means, but Has() should generally return true because the tombstone that we are writing has not been written yet, so there will be at least one missing tombstone for the value, unless we are re-adding a tombstone.